### DISCUSSION TITLE
Multiple versions of Java installed.

### DISCUSSION THREAD

**Discussion by j-antoniogarcia** [MinecraftServerControl/mscs#325] [Q&A] [answered]
Hello, I'm a complete noob when it comes to messing with Java and settings and such. I currently have two versions of Java on my PC, Java 8 and Java 18, with Java 18 being the default. Before I used MSCS, I would explicitly state in a .sh file the path to the java version I needed and would be able to run my server like normal. I am currently trying to mess with the mscs.properties file and thought I could do the same as before, using the mscs-server-command, and explicitly state the path to the Java I needed for the server version. However, after trying to run it, it keeps telling me "Java not found". I know the simple solution would probably be to just switch the default Java version, but I rather be able to run the servers seamlessly without changing it. How do I achieve this in mscs? Is there a config file or different settings I must change? Any response is appreciated, thank you.

**Comment by zanix** [comment, marked as answer]
In the `mscs-server-command` setting, you want to replace `$JAVA` with the _full_ path to the version you want to use.

**Reply by j-antoniogarcia** [reply]
Yeah I did that originally. Just now I replaced all variables with the values I wanted explicitly and now it's telling me that it couldn't retrieve the process ID, despite me using the correct Java version for the server.

**Reply by sandain** [reply]
What is the content of the `console.out` file for this server? If used in the standard way, you can find this file in`/opt/mscs/worlds/<world name>/console.out`.  This should tell you what went wrong.

**Reply by zanix** [reply]
Maybe `getJavaPID` isn't working right with a different path?
https://github.com/MinecraftServerControl/mscs/blob/main/msctl#L333

**Reply by j-antoniogarcia** [reply]
I checked the console and it said that permission was denied to the Java directory, which duh I should've figured out. I tried making it so that specifically the minecraft user would have execution and read permissions, but before that did a lot of trial and error methods. Safe to say I may have borked the program because now I can't run a single mscs command. When I don't use sudo, it says permission denied, and when I do use sudo, it says that once again, Java was not found, despite me not really moving anything around. It's saying this on ANY command now, not just starting a server. I don't know if ima keep messing with it or just leave it alone and come back to it another day. Thankfully I have backups of most everything so I should be good.
Edit: specifically, console.out said this: "sh: 2: /home/mintplays/Documents/jdk8u362-b09-jre/bin/java: Permission denied"
Edit 2: Ok I am no longer having weird issues with typing any command. Still can't start my world though lol.

**Reply by sandain** [reply]
Your minecraft user is going to need read access to everything in the `/home/mintplays/Documents/jdk8u362-b09-jre` folder.  I would suggest moving this folder out of your home folder and setting permissions for others:
```
mkdir -p /opt/java
mv /home/mintplays/Documents/jdk8u362-b09-jre /opt/java/jdk8u362-b09-jre
chmod -R o+r /opt/java/jdk8u362-b09-jre
chmod o+x /opt/java/jdk8u362-b09-jre/bin/*
```

Hopefully that helps.

**Reply by j-antoniogarcia** [reply]
I have moved everything into the opt folder. I've tried my best to update all paths using update-alternatives, and have confirmed java -version works as expected. However, I am once again getting a "ERROR: Java not found!" anytime I run a mscs command. I honestly don't know how I fixed it last time, but this time I'm stumped. I made sure permissions are set up properly, even using the setfacl to double check. I can't really progress from this point now.

**Reply by sandain** [reply]
Please post your `mscs.properties` file for the failing world.

**Reply by j-antoniogarcia** [reply]
```
mscs-enabled=true
mscs-server-version=1.16.5
mscs-client-version=1.16.5
mscs-server-command=/opt/java/jdk-8/bin/java -server -Xms$INITIAL_MEMORY -Xmx$MAXIMUM_MEMORY $JVM_ARGS -jar /opt/mscs/server/forge-1.16.5-36.2.35.jar nogui
```
The server command is the exact one I used before mscs. I have not been able to even attempt to start a world because just typing `mscs list` gives me the Java not found error. I have attempted removing the $INITIAL_MEMORY and $MAXIMUM_MEMORY btw.

**Reply by zanix** [reply]
Do you have a default Java? The script does try to run `which java` to get the default Java path.
EDIT: I just saw you tried `java -version`, can you try:

```shell
sudo -u minecraft which java
```

**Reply by sandain** [reply]
I'm not sure where the issue lies.  I would try switching to the minecraft user 
```
sudo su -s /bin/bash minecraft
```
then test java as the minecraft user
```
/opt/java/jdk-8/bin/java
```

If the minecraft user can access java, then the problem is elsewhere. I'm just not sure where

**Reply by j-antoniogarcia** [reply]
Ok so I made some progress, after just changing ONE thing, giving the bin folder itself o+x permissions, I can now use mscs commands again. Right now, I am in the bin directory as the minecraft user and attempting to do `./java -version` but now I am seeing this error: `./java: error while loading shared libraries: libjli.so: cannot open shared object file: No such file or directory`. I really appreciate y'alls help btw, I'm going to keep messing around and hopefully not mess it up more.

**Reply by j-antoniogarcia** [reply]
It might be fixed??? After being utterly defeated for a second, it seems to finally go live. I say that because after running the start command, literally seconds after it kicks this error at me: `find: Failed to restore initial working directory: /home/mintplays: Permission denied` No clue why it wanted to access my home directory, unless it needs to in order to access the other mscs files I had seen before. I tried changing something, then tried starting it again but it said "world already running" and was confused for a second. I confirm in status that it said the server was up, I launch minecraft and sure enough I can join the server. It seems the fix was just chmod o+x to all of the lib folder too.

**Reply by sandain** [reply]
Good. I'm glad you got it working.

**Reply by j-antoniogarcia** [reply]
I really appreciate y'alls help, I would definitely not have figured anything out without y'all. I've gone and tested the rest of the worlds I had imported and 3/4 are working. Last one doesn't work due to it having a very particular and complex start.sh file and I can't really get it to work with mscs. It's whatever though, overall everything else works as expected. Thank you all.
